airbyte-integrations/connectors/source-posthog/source_posthog/streams.py

```diff
@@ -0,0 +1,312 @@
+from abc import ABC, abstractmethod
+from typing import Any, Iterable, Mapping, MutableMapping, Optional, Tuple, Union
+
+import requests
+from airbyte_cdk.sources.streams.http import HttpStream
+
+
+class PosthogStream(HttpStream, ABC):
+    url_base = "https://app.posthog.com/api/"
+    primary_key = "id"
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        resp_json = response.json()
+        if resp_json.get("next"):
+            next_query_string = urllib.parse.urlsplit(resp_json["next"]).query
+            params = dict(urllib.parse.parse_qsl(next_query_string))
+            if params:
+                return params
+
+    def request_headers(self, **kwargs) -> Mapping[str, Any]:
+        return {"Content-Type": "application/json", "User-Agent": "posthog-python/1.4.0"}
+
+    def parse_response(self, response: requests.Response, stream_state: Mapping[str, Any], **kwargs) -> Iterable[Mapping]:
+        response_json = response.json()
+        yield from response_json.get(self.data_field)
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+
+        params = {}
+        if next_page_token:
+            params.update(next_page_token)
+        return params
+
+    @property
+    @abstractmethod
+    def data_field(self) -> str:
+        """the responce entry that contains useful data"""
+
+
+class IncrementalPosthogStream(PosthogStream, ABC):
+    state_checkpoint_interval = math.inf
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.reversed_pagination = {"is_completed": False, "latest_response_state": {}, "is_new_page_available": False, "upgrade_to": {}}
+
+    @property
+    @abstractmethod
+    def cursor_field(self) -> str:
+        """
+        Defining a cursor field indicates that a stream is incremental, so any incremental stream must extend this class
+        and define a cursor field.
+        """
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        resp_json = response.json()
+        if self.reversed_pagination["is_completed"]:
+            return {}
+        if "next" in resp_json and resp_json["next"]:
+            next_query_string = urllib.parse.urlsplit(resp_json["next"]).query
+            params = dict(urllib.parse.parse_qsl(next_query_string))
+            if params:
+                return params
+
+    def get_updated_state(
+        self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]
+    ) -> Union[Mapping[str, Any], Tuple[Any, Mapping[str, Any]]]:
+        """
+        Return the latest state by comparing the cursor value in the latest record with the stream's most recent state object
+        and returning an updated state object.
+        """
+        current_stream_state = current_stream_state or {}
```

Comment:
It is too hard, are you sure, that we can`t simplify this?

Reply:
in get_updated state it is possible we have None state, usually at first record. I'll get AttributeError on .get() from NoneType in this case and will need to perform this action anyway

Reply:
I understand, I mean that could we simplify this method, not current string?